lib/mwUtil.js

```diff
@@ -301,14 +301,31 @@ mwUtil.createRelativeTitleRedirect = (path, req, options) => {
             .replace(pathBeforeTitle, '');
     const pathSuffixCount = (pathSuffix.match(/\//g) || []).length;
     const backString = Array.apply(null, { length: pathSuffixCount }).map(() => '../').join('');
+    let location;
     if (!options.dropPathAfterTitle) {
         const pathPatternAfterTitle = path.substring(path.indexOf('{title}') - 1);
-        return backString
+        location = backString
             + new URI(pathPatternAfterTitle, newReqParams, true).toString().substr(1)
             + mwUtil.getQueryString(req);
     } else {
-        return backString + encodeURIComponent(newReqParams.title) + mwUtil.getQueryString(req);
+        location = backString + encodeURIComponent(newReqParams.title) + mwUtil.getQueryString(req);
     }
+
+    if (mwUtil.isSelfRedirect(req, location)) {
+        location = mwUtil.addQueryString(location, { redirect: false });
+    }
+
+    return location;
+};
+
+/**
+  * Checks if the request is a CORS request
+  *
+  * @param {Object} req The request to check
+  * @return {boolean} Whether the request is CORS or not
+  */
+mwUtil.isCrossOrigin = (req) => {
+    return req && req.headers && req.headers.origin && req.headers.origin !== req.params.domain;
```

Comment:
Minor: In some situations it might be more desirable to fail if no `req` or `req.headers` is passed in. Returning `false` here might instead cover up a bug. Perhaps we could just remove those extra checks?